// astro@2.10.7/dist/core/dev/container.js
import nodeFs from "node:fs";
import * as vite from "vite";
import { injectImageEndpoint } from "../../assets/internal.js";
import {
  runHookConfigDone,
  runHookConfigSetup,
  runHookServerDone,
  runHookServerStart
} from "../../integrations/index.js";
import { createVite } from "../create-vite.js";
import { apply as applyPolyfill } from "../polyfill.js";
async function createContainer({
  isRestart = false,
  logging,
  inlineConfig,
  settings,
  fs = nodeFs
}) {
  applyPolyfill();
  settings = await runHookConfigSetup({
    settings,
    command: "dev",
    logging,
    isRestart
  });
  if (settings.config.experimental.assets) {
    settings = injectImageEndpoint(settings);
  }
  const { host, headers, open } = settings.config.server;
  const rendererClientEntries = settings.renderers.map((r) => r.clientEntrypoint).filter(Boolean);
  const viteConfig = await createVite(
    {
      mode: "development",
      server: { host, headers, open },
      optimizeDeps: {
        include: rendererClientEntries
      }
    },
    { settings, logging, mode: "dev", command: "dev", fs }
  );
  await runHookConfigDone({ settings, logging });
  const viteServer = await vite.createServer(viteConfig);
  const container = {
    inlineConfig: inlineConfig ?? {},
    fs,
    logging,
    restartInFlight: false,
    settings,
    viteServer,
    handle(req, res) {
      viteServer.middlewares.handle(req, res, Function.prototype);
    },
    // TODO deprecate and remove
    close() {
      return closeContainer(container);
    }
  };
  return container;
}
async function closeContainer({ viteServer, settings, logging }) {
  await viteServer.close();
  await runHookServerDone({
    config: settings.config,
    logging
  });
}
async function startContainer({
  settings,
  viteServer,
  logging
}) {
  const { port } = settings.config.server;
  await viteServer.listen(port);
  const devServerAddressInfo = viteServer.httpServer.address();
  await runHookServerStart({
    config: settings.config,
    address: devServerAddressInfo,
    logging
  });
  return devServerAddressInfo;
}
export {
  createContainer,
  startContainer
};
