// astro@2.10.7/dist/core/preview/index.js
import { createRequire } from "node:module";
import { fileURLToPath, pathToFileURL } from "node:url";
import { telemetry } from "../../events/index.js";
import { eventCliSession } from "../../events/session.js";
import { runHookConfigDone, runHookConfigSetup } from "../../integrations/index.js";
import { resolveConfig } from "../config/config.js";
import { createNodeLogging } from "../config/logging.js";
import { createSettings } from "../config/settings.js";
import createStaticPreviewServer from "./static-preview-server.js";
import { getResolvedHostForHttpServer } from "./util.js";
async function preview(inlineConfig) {
  const logging = createNodeLogging(inlineConfig);
  const { userConfig, astroConfig } = await resolveConfig(inlineConfig ?? {}, "preview");
  telemetry.record(eventCliSession("preview", userConfig));
  const _settings = createSettings(astroConfig, fileURLToPath(astroConfig.root));
  const settings = await runHookConfigSetup({
    settings: _settings,
    command: "preview",
    logging
  });
  await runHookConfigDone({ settings, logging });
  if (settings.config.output === "static") {
    const server2 = await createStaticPreviewServer(settings, logging);
    return server2;
  }
  if (!settings.adapter) {
    throw new Error(`[preview] No adapter found.`);
  }
  if (!settings.adapter.previewEntrypoint) {
    throw new Error(
      `[preview] The ${settings.adapter.name} adapter does not support the preview command.`
    );
  }
  const require2 = createRequire(settings.config.root);
  const previewEntrypointUrl = pathToFileURL(
    require2.resolve(settings.adapter.previewEntrypoint)
  ).href;
  const previewModule = await import(previewEntrypointUrl);
  if (typeof previewModule.default !== "function") {
    throw new Error(`[preview] ${settings.adapter.name} cannot preview your app.`);
  }
  const server = await previewModule.default({
    outDir: settings.config.outDir,
    client: settings.config.build.client,
    serverEntrypoint: new URL(settings.config.build.serverEntry, settings.config.build.server),
    host: getResolvedHostForHttpServer(settings.config.server.host),
    port: settings.config.server.port,
    base: settings.config.base
  });
  return server;
}
export {
  preview as default
};
