// astro@2.10.7/dist/core/errors/errors.js
import { codeFrame } from "./printer.js";
function isAstroError(e) {
  return e instanceof Error && e.type === "AstroError";
}
class AstroError extends Error {
  constructor(props, ...params) {
    super(...params);
    this.type = "AstroError";
    const { name, title, message, stack, location, hint, frame } = props;
    this.title = title;
    this.name = name;
    if (message)
      this.message = message;
    this.stack = stack ? stack : this.stack;
    this.loc = location;
    this.hint = hint;
    this.frame = frame;
  }
  setLocation(location) {
    this.loc = location;
  }
  setName(name) {
    this.name = name;
  }
  setMessage(message) {
    this.message = message;
  }
  setHint(hint) {
    this.hint = hint;
  }
  setFrame(source, location) {
    this.frame = codeFrame(source, location);
  }
  static is(err) {
    return err.type === "AstroError";
  }
}
class CompilerError extends AstroError {
  constructor(props, ...params) {
    super(props, ...params);
    this.type = "CompilerError";
  }
  static is(err) {
    return err.type === "CompilerError";
  }
}
class CSSError extends AstroError {
  constructor() {
    super(...arguments);
    this.type = "CSSError";
  }
  static is(err) {
    return err.type === "CSSError";
  }
}
class MarkdownError extends AstroError {
  constructor() {
    super(...arguments);
    this.type = "MarkdownError";
  }
  static is(err) {
    return err.type === "MarkdownError";
  }
}
class InternalError extends AstroError {
  constructor() {
    super(...arguments);
    this.type = "InternalError";
  }
  static is(err) {
    return err.type === "InternalError";
  }
}
class AggregateError extends AstroError {
  // Despite being a collection of errors, AggregateError still needs to have a main error attached to it
  // This is because Vite expects every thrown errors handled during HMR to be, well, Error and have a message
  constructor(props, ...params) {
    super(props, ...params);
    this.type = "AggregateError";
    this.errors = props.errors;
  }
  static is(err) {
    return err.type === "AggregateError";
  }
}
const astroConfigZodErrors = /* @__PURE__ */ new WeakSet();
function isAstroConfigZodError(error) {
  return astroConfigZodErrors.has(error);
}
function trackAstroConfigZodError(error) {
  astroConfigZodErrors.add(error);
}
export {
  AggregateError,
  AstroError,
  CSSError,
  CompilerError,
  InternalError,
  MarkdownError,
  isAstroConfigZodError,
  isAstroError,
  trackAstroConfigZodError
};
